cordova-plugin-outline/android/java/org/outline/vpn/VpnTunnelService.java

```diff
@@ -430,74 +406,60 @@ private void stopNetworkConnectivityMonitor() {
     }
   }
 
-  // Broadcasts
-
-  /* Broadcast VPN start. |errorCode| determines whether the VPN was started successfully, or
-   * specifies the error condition. */
-  private void broadcastVpnStart(OutlinePlugin.ErrorCode errorCode) {
-    Intent vpnStart = new Intent(OutlinePlugin.Action.START.value);
-    vpnStart.putExtra(OutlinePlugin.IntentExtra.ERROR_CODE.value, errorCode.value);
-    dispatchBroadcast(vpnStart);
-  }
-
-  /* Broadcast VPN stop. */
-  private void broadcastVpnStop() {
-    Intent vpnStop = new Intent(OutlinePlugin.Action.STOP.value);
-    vpnStop.putExtra(OutlinePlugin.IntentExtra.ERROR_CODE.value,
-                     OutlinePlugin.ErrorCode.NO_ERROR.value);
-    dispatchBroadcast(vpnStop);
-  }
-
   /* Broadcast change in the VPN connectivity. */
   private void broadcastVpnConnectivityChange(OutlinePlugin.TunnelStatus status) {
     Intent statusChange = new Intent(OutlinePlugin.Action.ON_STATUS_CHANGE.value);
-    statusChange.putExtra(OutlinePlugin.IntentExtra.PAYLOAD.value, status.value);
+    statusChange.addCategory(getPackageName());
+    statusChange.putExtra(OutlinePlugin.MessageData.PAYLOAD.value, status.value);
     statusChange.putExtra(
-        OutlinePlugin.IntentExtra.ERROR_CODE.value, OutlinePlugin.ErrorCode.NO_ERROR.value);
-    dispatchBroadcast(statusChange);
-  }
-
-  private void dispatchBroadcast(Intent broadcast) {
-    if (activeTunnelId != null) {
-      broadcast.putExtra(OutlinePlugin.IntentExtra.TUNNEL_ID.value, activeTunnelId);
-    }
-    LocalBroadcastManager.getInstance(this).sendBroadcast(broadcast);
+        OutlinePlugin.MessageData.ERROR_CODE.value, OutlinePlugin.ErrorCode.NO_ERROR.value);
+    statusChange.putExtra(OutlinePlugin.MessageData.TUNNEL_ID.value, tunnelConfig.id);
+    sendBroadcast(statusChange);
   }
 
   // Autostart
 
-  private void startLastSuccessfulTunnelOrExit() {
+  private void startLastSuccessfulTunnel() {
     LOG.info("Received an auto-connect request, loading last successful tunnel.");
     JSONObject tunnel = tunnelStore.load();
     if (tunnel == null) {
       LOG.info("Last successful tunnel not found. User not connected at shutdown/install.");
-      stopSelf();
       return;
     }
     if (VpnTunnelService.prepare(VpnTunnelService.this) != null) {
       // We cannot prepare the VPN when running as a background service, as it requires UI.
       LOG.warning("VPN not prepared, aborting auto-connect.");
-      stopSelf();
       return;
     }
     try {
-      final JSONObject config = tunnel.getJSONObject(TUNNEL_CONFIG_KEY);
+      final TunnelConfig config = new TunnelConfig();
```

Comment:
This code to convert from JSON to TunnelConfig is duplicated in OutlinePlugin.  They should be able to share this as a public static method on VpnTunnelService.